keps/sig-scheduling/5075-dra-consumable-capacity/README.md

```diff
@@ -0,0 +1,1150 @@
+	// +optional
+	// +listType=map
+	// +listMapKey=driver
+	// +listMapKey=device
+	// +listMapKey=pool
+	// +listMapKey=shareUID
+	// +featureGate=DRAResourceClaimDeviceStatus
+	Devices []AllocatedDeviceStatus `json:"devices,omitempty" protobuf:"bytes,4,opt,name=devices"`
+}
+
+type DeviceRequestAllocationResult struct {
+  ...
+
+   // ShareUID uniquely identifies the specific allocation result of the shareable device.
+   // Set only when allocation is on a shareable device.
+   //
+   // +optional
+   // +featureGate=DRAConsumableCapacity
+   ShareUID *types.UID
+
+  // Alternatively, SharedAllocationIndex could have been used as a reference to the allocation result.
+  // However, the index may become outdated if the allocation is reallocated (should that ever be supported),
+  // making it less reliable than ShareUID.
+
+  // ConsumedCapacities tracks the amount of capacity consumed per device as part of the claim request.
+  // The consumed amount may differ from the requested amount: it is rounded up to the nearest valid
+  // value based on the device’s sharing policy.
+  //
+  // The total consumed capacity for each device must not exceed its available capacity.
+  //
+  // This field references only consumable capacities of a shareable device and is empty when there are none.
+  //
+  // +optional
+  // +featureGate=DRAConsumableCapacity
+   ConsumedCapacities map[QualifiedName]resource.Quantity
+}
+```
+
+### Scheduling enhancement
+- When the scheduler invokes the `Allocate` function in the allocator, 
+  the total allocated capacity is calculated by aggregating the ConsumedCapacities from all resource claims's `DeviceRequestAllocationResult` that have already been allocated.
+- Before allocation proceeds, existing selection criteria (defined by `alloc.isSelectable`) are evaluated. 
+  These include the class selector and request selector.
+- A new `device.shareable` key is introduced in the CEL selector,
+  enabling policies and constraints to recognize whether a device supports shareable allocation.
+- If a device is considered selectable, the `CmpRequestOverCapacity` function is invoked to verify 
+  whether the consumed capacity would exceed the device's remaining capacity. 
+  The remaining capacity is calculated based on the sum of already allocated and currently allocating capacities.
+  - consumed capacity is derived from the requested amount specified in the resource claim, adjusted by the device’s capacity sharing policy, if defined.
+  - This value may differ from the originally requested amount—it is rounded up to the nearest valid capacity according to the policy (e.g., using Minimum + ⌈(Requested - Minimum)/ChunkSize⌉ × ChunkSize logic).
+- If the device has enough remaining capacity to satisfy the consumed amount, constraint checks are applied. 
+  In addition to the existing MatchAttribute, this proposal introduces a new constraint: `DistinctAttribute`, which ensures attribute uniqueness across allocated devices.
```

Comment:
Do we need this feature in this KEP? I mean, is it necessary for the use cases? If so that is fine, but if not we could potentially do it in a separate KEP just to reduce scope. I am OK either way.

Reply:
Wouldn't this be possible to achieve the same with the [KEP-5254 - Constraints with CEL](https://github.com/kubernetes/enhancements/issues/5254)? CEL checking if the devices have the same/distinct attributes would be more generic than `DistinctAttribute`

Reply:
Yes, but that requires writing CEL. I think it's nice to have the symmetry with matchAttributes without writing CEL.

There are also two different ideas we have discussed with respect to #5254.

1) I think the one in the KEP is a single CEL expression evaluated over the entire set of devices. This could achieve the distinct attributes but iterating over a set is pretty advanced CEL.
2) A "matchExpression" which works just like matchAttribute but instead of reading an attribute, evaluates a per-device expression. The results of that expression would then have to be the same for every device in the set.

If we do 2), then we can also do a `distinctExpression`. Though I believe 1) is the thing we are actually doing in that KEP.